### DISCUSSION TITLE
How to create a Graphviz object (not a pygraphviz or pydot object) from a networkx object?

### DISCUSSION THREAD

**Discussion by brando90** [networkx/networkx#4784] [Q&A]
I saw the docs: https://networkx.org/documentation/stable/reference/drawing.html#module-networkx.drawing.layout but it doesn't have that option. I found it puzzling since python graphviz has the largest user base (~19k users vs 15k for pydot and ~4k for pygraphviz). So:

1. why is that so?
2. how do I create a graphviz object?
3. what is the recommended way to visualize networkx graphs? why not graphviz but the other two are supported?

**Comment by rossbar** [comment]
I'm not sure I understand the question... are you referring to the [gv3.python scripting API provided by GraphViz](https://graphviz.org/documentation/)? Here's some (hopefully relevant) background:

GraphViz provides utilities that AIUI are typically used via command line interface to modify dot files on disk, so to "use GraphViz from Python", you need some sort of interface to do so. `pydot` and `pygraphviz` are two such packages providing these interfaces. The primary interface from pygraphviz is the [AGraph class](https://pygraphviz.github.io/documentation/stable/reference/agraph.html), which is similar to the `Graph` interface familiar to NetworkX users. `pydot` provides it's [own interface](https://github.com/pydot/pydot#about) for creating/loading/modifying DOT files.

**Reply by brando90** [reply]
sorry I missed a word, this is what I meant to ask (title corrected): 

> How to create a Graphviz object (not a pygraphviz or pydot object) from a networkx object?

**Comment by dschult** [comment]
Generically and historically, Graphviz works with text files that have the `dot` format. So there isn't such a thing as a `graphviz object`.  The Graphviz library does have classes like the older `agraph` and now `cgraph`. But they are created with the goal of reading and writing dot files. `pygraphviz` uses the `cgraph` library to read and write dot files which get processed by Graphviz.

Your question is confusing because you use the term `graphviz` which to many of us refers to the C library Graphviz. But there is a python package called graphviz as well. Perhaps you are referring to the python package called `graphviz`, or often `python-graphviz`.  If so, the reason we provide support for pydot and pygraphviz but not to python-graphviz is because no one has written an interface to python-graphviz.  Since we already have two ways to interact with dot files someone would need a reason to implement a python-graphviz read-write module. Also, pydot and pygraphviz started in 2004 and the python-graphviz project started in 2014. Basically, it would need someone to take an interest to create the interface in networkx.

**Reply by brando90** [reply]
sorry I missed a word, this is what I meant to ask (title corrected): 

> How to create a Graphviz object (not a pygraphviz or pydot object) from a networkx object?

Just wanted to make the clarification explicit.
